packages/eslint-plugin/src/rules/method-signature-style.ts

```diff
@@ -0,0 +1,118 @@
+import {
+  AST_NODE_TYPES,
+  TSESTree,
+} from '@typescript-eslint/experimental-utils';
+import * as util from '../util';
+
+export type Options = ['property' | 'method'];
+
+export type MessageId = 'errorMethod' | 'errorProperty';
+
+export default util.createRule<Options, MessageId>({
+  name: 'method-signature-style',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Enforces using a particular method signature syntax.',
+      category: 'Best Practices',
+      recommended: false,
+    },
+    fixable: 'code',
+    messages: {
+      errorMethod:
+        'Shorthand method signature is forbidden. Use a function property instead.',
+      errorProperty:
+        'Function property signature is forbidden. Use a method shorthand instead.',
+    },
+    schema: [
+      {
+        enum: ['property', 'method'],
+      },
+    ],
+  },
+  defaultOptions: ['property'],
+
+  create(context, [mode]) {
+    const sourceCode = context.getSourceCode();
+
+    function getMethodKey(
+      node: TSESTree.TSMethodSignature | TSESTree.TSPropertySignature,
+    ): string {
+      let key = sourceCode.getText(node.key);
+      if (node.computed) {
+        key = `[${key}]`;
+      }
+      if (node.optional) {
+        key = `${key}?`;
+      }
+      if (node.readonly) {
+        key = `readonly ${key}`;
+      }
+      return key;
+    }
+
+    function getMethodParams(
+      node: TSESTree.TSMethodSignature | TSESTree.TSFunctionType,
+    ): string {
+      let params = node.params.map(node => sourceCode.getText(node)).join(', ');
+      params = `(${params})`;
+      if (node.typeParameters != null) {
+        const typeParams = sourceCode.getText(node.typeParameters);
+        params = `${typeParams}${params}`;
+      }
+      return params;
+    }
```

Comment:
This will remove all comments inside the type.
For the most part I don't have problems with a fixer removing comments in weird places if it makes the fixer simpler and easier to maintain.

But one place we should accept and support comments is around parameters, eg:
```
method(/* leading comment */arg: string): void
```

So I would suggest converting this to something like:
- grab first open paren token
- grab last end paren token
- grab source code between the two

That way we preserve the "expected" comments

Reply:
I just did that, but it only works if there's at least one argument. If the argument list is `(/* comment */)` then the comment is lost. The problem is that there's no way to find the paren tokens because the method node's "params" property is just an empty array. This is contrary to "typeParams" which is a separate AST node so I can just `getText` it even if it's empty.